fuel-core/src/schema/contract.rs

```diff
@@ -113,4 +124,87 @@ impl ContractBalanceQuery {
             asset_id,
         })
     }
+
+    async fn contract_balances(
+        &self,
+        ctx: &Context<'_>,
+        filter: ContractBalanceFilterInput,
+        first: Option<i32>,
+        after: Option<String>,
+        last: Option<i32>,
+        before: Option<String>,
+    ) -> async_graphql::Result<Connection<AssetId, ContractBalance, EmptyFields, EmptyFields>> {
+        let db = ctx.data_unchecked::<Database>().clone();
+
+        query(
+            after,
+            before,
+            first,
+            last,
+            |after: Option<AssetId>, before: Option<AssetId>, first, last| async move {
+                // Calculate direction of which to iterate through rocksdb
+                let (records_to_fetch, direction) = if let Some(first) = first {
+                    (first, IterDirection::Forward)
+                } else if let Some(last) = last {
+                    (last, IterDirection::Reverse)
+                } else {
+                    (0, IterDirection::Forward)
+                };
+
+                if (first.is_some() && before.is_some()) || (after.is_some() && before.is_some()) {
+                    return Err(anyhow!("Wrong argument combination"));
+                }
+
+                let after = after.map(fuel_tx::AssetId::from);
+                let before = before.map(fuel_tx::AssetId::from);
+
+                let start = if direction == IterDirection::Forward {
+                    after
+                } else {
+                    before
+                };
+
+                let mut balances_iter =
+                    db.contract_balances(filter.contract.into(), start, Some(direction));
+
+                let mut started = None;
+                if start.is_some() {
+                    started = balances_iter.next();
+                }
+
+                let balances: Result<Vec<ContractBalance>, KvStoreError> = balances_iter
+                    .take(records_to_fetch + 1)
+                    .map(|balance| {
+                        let balance = balance?;
+
+                        Ok(ContractBalance {
+                            contract: filter.contract.into(),
+                            amount: balance.1,
+                            asset_id: balance.0,
+                        })
+                    })
+                    .collect();
+                let mut balances = balances?;
```

Comment:
[nit] you could use the turbofish syntax to combine these lines:

```suggestion
                let balances  = balances_iter
                    .take(records_to_fetch + 1)
                    .map(|balance| {
                        let balance = balance?;

                        Ok(ContractBalance {
                            contract: filter.contract.into(),
                            amount: balance.1,
                            asset_id: balance.0,
                        })
                    })
                    .collect::<Result<Vec<ContractBalance>, KvStoreError>>()?;
```